### ISSUE TITLE
Bug: Text in editor is constantly moving

### ISSUE THREAD

**Issue by sanderdatema** [lmparppei/Beat#138]
Somehow the text is moving a few pixels back and forth constantly. This GIF shows what I mean. This shows with any new document, just type a bit of text. It's ticking like a clock, about every half a second.

![Blinking_Example](https://github.com/lmparppei/Beat/assets/176252/6c364aea-992b-4759-b836-a9670c33788c)

MacBook M1 Pro, Sonoma 14.0. Beat version 1.997.0, build 133241. No plugins.

**Comment by lmparppei** [comment]
_Super_ weird. Could you provide some more info:
- Is it happening even when not typing?
- Is **Hide Fountain Markup** on?
- What sort of screen are you using, laptop or external?
- In your screen settings, are you using default setting for resolution or scaled?

**Comment by lmparppei** [comment]
**Update:** This is very likely a TextKit 1 compatibility bug in macOS Sonoma, as they are pushing towards TextKit 2, which is still quite buggy. This should be reported to Apple, but they are not very keen on fixing TextKit 1 issues anymore. 

You can try adjusting the zoom level and see if it makes a difference.

Beat relies heavily on TextKit 1 features, and isn't migrating anytime soon, so we'll have to hope Apple fixes rendering issues. I'll try to figure out a workaround.

**Comment by sanderdatema** [comment]
Thanks for the quick reply. Here are my answers:
- Yes, it happens even when not typing.
- It makes no difference if it's on or not.
- **Interesting, it only happens on an external monitor. It's perfect on the internal monitor of my MacBook.** The wiggle on the external monitor is only on the active window (when having multiple documents open in Beat)
- On my internal screen all resolutions were fine (see previous point)

**Comment by lmparppei** [comment]
> EDIT: I was wrong, it also happens on the internal monitor, any resolution, but only on the active window.

I'm pretty sure this is a TextKit 2 issue, fonts are rendered wrong when caret is drawn. Does zoom level (you can adjust it by pressing cmd +/-) affect the wiggle in any way?

**Comment by sanderdatema** [comment]
> Does zoom level affect the wiggle in any way?

Nope.

**Comment by lmparppei** [comment]
Does this build make a difference? It changes how the caret is drawn.

https://www.dropbox.com/s/f9ywg41kt4cswgv/Beat%201.997.0.dmg?dl=1

**Comment by sanderdatema** [comment]
I'm afraid not. The problem is appearing in the same way. Please let me know if there's anything I can do on my side to help you debug this.

If this is caused by the TextKit 1 issue you mentioned before, then the solution might be out of our reach.

**Comment by lmparppei** [comment]
Can you try it with a section line? Does that wiggle as well, or stay in place?

<img width="482" alt="image" src="https://github.com/lmparppei/Beat/assets/20226704/4c261d8f-9572-42e6-b329-8876b1ac91ba">

**Comment by ftolsson** [comment]
Just to add my two cents: It can't squarely be a Sonoma issue since I'm running Beat perfectly on Sonoma (on an M2) with no wiggliness.
@sanderdatema: looking at the screenshot, I'm curious why the page shadow is flat for you -- for me, it's a smooth gradient, se below. Is this a compression thing from the screencast or might there be some screen settings at play? (I also tried to look for any way to set the cursor size in case that might affect something but couldn't immediately find it -- have you made any changes to any accessibility settings?) 

<img width="1113" alt="image" src="https://github.com/lmparppei/Beat/assets/25481285/be2abd3f-9518-4364-bf4a-af209175f57a">

**Comment by ftolsson** [comment]
I stand corrected. Zooming in *heavily* on the screen I get the same -- it wasn't visible to my naked eye on my high res monitor (did the original video come from a non-retina screen?)

As far as I can tell, it seems the text jumps up by almost exactly half a screen pixel when the cursor is visible. 

Can't tell if this happens on any other system of course, nor can I tell for sure if this also happens for someone who has never set the size of the pointer up to max in the accessibility settings, because I did to try to provoke the wiggle (oh the things we do to procrastinate from writing). 

The wiggle now remains when resetting the pointer size to normal so it may well be unrelated. It was just that I didn't magnify the screen, and so did not discover it, until after I tried that.

Hope any of this helps.

![blinkage](https://github.com/lmparppei/Beat/assets/25481285/25e89e88-bb9b-43e4-b308-4d48fd0a8762)

**Comment by lmparppei** [comment]
It's not so much that the text actually moves anywhere, but gets anti-aliased differently.

**Comment by lmparppei** [comment]
Try this build:
https://www.dropbox.com/s/f9ywg41kt4cswgv/Beat%201.997.0b.dmg?dl=1

Changes made:
- caret drawing avoids integral values
- `wantsLayer` and `canDrawSubviewsIntoLayer` set `true`

**Comment by ftolsson** [comment]
Ah, wait. Bad testing. My fault. It **doesn't** happen in my old documents, just the **new one** I created to test stuff. 

So in a blank doc, it still moves/changes aliasing even in your new build. 
But in the middle of a page of an old doc it does not (and also did not before, so I was actually right when I initially said I hadn't seen it.)

Let's see if this depends on no pagination, no scroll, something else!
Aynways seems to quite isolated.
Sorry!

**Comment by ftolsson** [comment]
Preliminary report: It only happens if the cursor is on the last line of the script.

**Comment by ftolsson** [comment]
Yup. Which is to say, the last line of text. (you can have blank lines below, that doesn't help.)

So if one is the kind of writer who always writes forward, i.e. has no lines below one's caret, it happens all the time.
But if one, like me, will more often than not have pre-planned stuff below, one never sees it.

You know your architecture better than I of course, but could it be that the cursor somehow affects the height of the complete bounding box when it blinks at the absolute bottom and thus changes the height of the text mass?

**Comment by lmparppei** [comment]
> could it be that the cursor somehow affects the height of the complete bounding box when it blinks at the absolute bottom and thus changes the height of the text mass?

Not really – this should be determined by macOS components. Let's try what happens if I replace my whole custom layout manager. **This build is just for testing and breaks a lot of stuff**:
https://www.dropbox.com/s/nxlraugi6gucwt4/Beat%201.997.0c.dmg?dl=1

**Comment by ftolsson** [comment]
Still blinking. So not your custom manager's fault!

**Comment by sanderdatema** [comment]
@lmparppei Yep, still the same.

@ftolsson I think thew shadow was cut off because I cropped it quite a lot. I used LICEcap to record the screencast.

**Comment by ftolsson** [comment]
For what it's worth — and you probably know this but still — your caret is slightly different from the other carets of macOS Sonoma.
Most noteworthy (if at all) is the fact that it's positioned some slight subpixel value offset from its closest integer, whereas the system cursor is always distinctly on-pixel. If that matters. Which it probably doesn't but I felt I should mention it.

<img width="2143" alt="image" src="https://github.com/lmparppei/Beat/assets/25481285/e88af22f-6add-41ba-a0f2-7719ac7e7919">

**Comment by lmparppei** [comment]
I'm not sure if this build does anything, I changed some of the caret code once again, but I doubt it as the earlier change to caret rendering didn't have any effect. TextKit 2 behavior is documented pretty badly, and this is just stabbing in the dark, as I'm still avoiding Sonoma because of my Intel Mac.
https://www.dropbox.com/s/jo45js56rgv0vpw/Beat%201.997.0d.dmg?dl=1

**Comment by lmparppei** [comment]
Oh, @ftolsson, if you are running Beat in Xcode, do you get a log message about TextKit switching to compatibility mode?

**Comment by sanderdatema** [comment]
The latest build doesn't change anything indeed.

**Comment by ftolsson** [comment]
You mean the previous builds? (Because I can't run these built ones from there can I?)

There are a couple of warnings but this one felt potentially relevant:

<img width="751" alt="image" src="https://github.com/lmparppei/Beat/assets/25481285/9285d14b-f025-4a82-a1f8-4123a6214bc2">

**Comment by lmparppei** [comment]
What the hell is that `NSRemoteView` message? It seems to relate to cursor somehow, but AFAIK `TUI` namespace is some Twitter UI framework. You are seeing this in Xcode log?

**Comment by ftolsson** [comment]
Ah, haha, sounds much more likely for sure. 
The metadata says it pertains to a library called ViewBridge if that this means anything to you:

```
Type: Fault | Timestamp: 2023-11-01 13:47:12.375626+01:00 | Library: ViewBridge | Subsystem: com.apple.ViewBridge | Category: miscellany | TID: 0x64a83a
```

**Comment by lmparppei** [comment]
Discussion on Apple forums kind of points to this being a TextKit bug on Sonoma. I'll update to Sonoma myself and see if there's nothing to do about it – and file a radar case for this.

**Comment by lmparppei** [comment]
Update:

This is a TextKit 1 compatibility bug in Sonoma, related to `NSTextView` scaling. Line fragment positions are not rounded correctly in non-fourth scale ratios. I've changed to default zoom level to `1.5` for the upcoming release, but anything else will cause trouble.

I'm working on a fix which uses scroll view magnification instead of text view scaling, but it will probably lead to further weirdness at first.

**Comment by lmparppei** [comment]
The moving might have been *somewhat* fixed by removing some compatibility code, but other TextKit 1 bugs still remain:
https://www.dropbox.com/s/yz5anorwwwwoojx/Beat%201.998.2.dmg?dl=1

Removing one piece of code might make the app crash now and then on 10.13, but let's hope not.

**Comment by lmparppei** [comment]
It seems like Apple has fixed this in 14.4.1!

I'm closing this unless new discoveries are made.
